### PR TITLE
internal/cmd/gocdk: add Cloud Run launcher

### LINKED ISSUE #1861: internal/cmd/gocdk: implement minimal launch command
The most minimal thing I can think of for this to do is to take a Docker image and get it to run with [Cloud Run](https://cloud.google.com/run/docs/). While the steps I outline below are specific to Cloud Run, the logic should be able to cleanly branch for other deployment environments.

The Docker image should already exist by the time `gocdk launch` runs (see #1807) and will be passed as a flag to `gocdk launch`, but the command will need to:

1. Gather the selected launcher from the biome's biome.json and the launch specifier from the biome's Terraform outputs (`output "launcher_specifier"`).
2. (Maybe) Push the Docker image to Google Container Registry.
3. Make the correct Cloud Run API calls to start up the service from the Docker image.

Later on, we can worry about propagating environment variables/configuration from Terraform outputs to the Cloud Run startup, but that can be additive.

### PR DESCRIPTION
**Pull request by zombiezen**
Fixes #1861

### CODE REVIEW

File: internal/cmd/gocdk/launch.go [hunk 1]
```diff
@@ -124,6 +130,22 @@ func newLauncher(ctx context.Context, pctx *processContext, launcherName string)
...
+		if err != nil {
+			return nil, xerrors.Errorf("prepare cloudrun launcher: %w", err)
+		}
+		httpClient, _ := gcp.NewHTTPClient(http.DefaultTransport, creds.TokenSource)
+		runService, err := cloudrun.NewService(ctx, option.WithHTTPClient(&httpClient.Client))
+		if err != nil {
+			return nil, xerrors.Errorf("prepare cloudrun launcher: %w", err)
+		}
+		return &cloudRunLauncher{
+			logger:    logger,
```
Comment (vangent) [inline internal/cmd/gocdk/launch.go]: Kind of same question as before: if `logger`, `dockerEnv`, and `dockerDir` are going to be needed by every launcher, why not pass them in `LaunchInput`?

File: internal/cmd/gocdk/launch.go [hunk 2]
```diff
@@ -187,6 +209,172 @@ func (local *localLauncher) Launch(ctx context.Context, input *LaunchInput) (*ur
...
+	dockerDir string
+}
+
+// Launch implements Launcher.Launch.
+func (crl *cloudRunLauncher) Launch(ctx context.Context, input *LaunchInput) (*url.URL, error) {
+	projectID := specifierStringValue(input.Specifier, "project_id")
+	location := specifierStringValue(input.Specifier, "location")
+	serviceName := specifierStringValue(input.Specifier, "service_name")
+	if projectID == "" || location == "" || serviceName == "" {
+		return nil, xerrors.New("cloud run launch: specifier does include project_id, location, and/or service_name")
```
Comment (vangent) [inline internal/cmd/gocdk/launch.go]: does -> missing ?

Comment (vangent) [reply internal/cmd/gocdk/launch.go]: And maybe be more specific than `specifier`; IIUC, this is the `launch_specifier` section in `outputs.tf`?

File: internal/cmd/gocdk/launch.go [hunk 3]
```diff
@@ -187,6 +209,172 @@ func (local *localLauncher) Launch(ctx context.Context, input *LaunchInput) (*ur
...
+	}
+
+	// Push to GCR if needed.
+	// TODO(light): Check for presence in Docker daemon and country-specific
+	// prefixes.
+	if strings.HasPrefix(input.DockerImage, "gcr.io/") {
```
Comment (vangent) [inline internal/cmd/gocdk/launch.go]: If it doesn't start with `gcr.io`, what does that mean? Is the rest of this going to work?

File: internal/cmd/gocdk/launch.go [hunk 4]
```diff
@@ -187,6 +209,172 @@ func (local *localLauncher) Launch(ctx context.Context, input *LaunchInput) (*ur
...
+			return nil, xerrors.Errorf("cloud run launch: %w", err)
+		}
+	}
+
+	// Launch on Cloud Run.
+	var env []*cloudrun.EnvVar
+	for i, inputVar := range input.Env {
+		eqIdx := strings.IndexByte(inputVar, '=')
+		if eqIdx == -1 {
+			return nil, xerrors.Errorf("cloud run launch: bad format for specifier env[%d] (got %q)", i, inputVar)
```
Comment (vangent) [inline internal/cmd/gocdk/launch.go]: Same here about being more specific about where this is coming from. And tell them what the format should be, not just that it's bad.

File: internal/cmd/gocdk/launch.go [hunk 5]
```diff
@@ -187,6 +209,172 @@ func (local *localLauncher) Launch(ctx context.Context, input *LaunchInput) (*ur
...
+		}
+		env = append(env, &cloudrun.EnvVar{
+			Name:  inputVar[:eqIdx],
```
Comment (vangent) [inline internal/cmd/gocdk/launch.go]: Nit: you could use `strings.Split` for this instead.

File: internal/cmd/gocdk/launch.go [hunk 6]
```diff
@@ -187,6 +209,172 @@ func (local *localLauncher) Launch(ctx context.Context, input *LaunchInput) (*ur
...
+			Value: inputVar[eqIdx+1:],
+		})
+	}
+	serviceMeta := &cloudrun.ObjectMeta{
+		Name:      serviceName,
+		Namespace: projectID,
+	}
+	serviceSpec := &cloudrun.ServiceSpec{
```
Comment (vangent) [inline internal/cmd/gocdk/launch.go]: Consider add a link here to where this is documented....

File: internal/cmd/gocdk/launch.go [hunk 7]
```diff
@@ -187,6 +209,172 @@ func (local *localLauncher) Launch(ctx context.Context, input *LaunchInput) (*ur
...
+		// Handle error below in the same way a create call is handled.
+		_, err = replaceCall.Context(ctx).Do()
+	} else if err == nil {
+		// Make publicly accessible.
+		policyCall := crl.client.Projects.Locations.Services.SetIamPolicy(serviceString, &cloudrun.SetIamPolicyRequest{
+			Policy: &cloudrun.Policy{
+				Bindings: []*cloudrun.Binding{
+					{
+						Role:    "roles/run.invoker",
+						Members: []string{"allUsers"},
```
Comment (vangent) [inline internal/cmd/gocdk/launch.go]: Should we make this be configured in `launch_specifier`?

File: internal/cmd/gocdk/launch.go [hunk 8]
```diff
@@ -187,6 +209,172 @@ func (local *localLauncher) Launch(ctx context.Context, input *LaunchInput) (*ur
...
+		// Already exists, add revision.
+		replaceCall := crl.client.Projects.Locations.Services.ReplaceService(serviceString, &cloudrun.Service{
+			ApiVersion: "serving.knative.dev/v1alpha1",
+			Kind:       "Service",
+			Metadata:   serviceMeta,
+			Spec:       serviceSpec,
+		})
+		// Handle error below in the same way a create call is handled.
+		_, err = replaceCall.Context(ctx).Do()
+	} else if err == nil {
```
Comment (vangent) [inline internal/cmd/gocdk/launch.go]: Nit: I found this error flow a bit hard to follow. How about flipping the if/else?

```
err := createCall.Do()
if err == nil {
  // set IAM
  err = policyCall.Do()
} else if err == StatusConflict {
  // Add revision
  err = replaceCall.Do()
}

if err != nil {
  // either the create failed, or the policy call failed, or the replace failed
  return err
}

...
```

File: internal/cmd/gocdk/launch.go [hunk 9]
```diff
@@ -187,6 +209,172 @@ func (local *localLauncher) Launch(ctx context.Context, input *LaunchInput) (*ur
...
+		// Make publicly accessible.
+		policyCall := crl.client.Projects.Locations.Services.SetIamPolicy(serviceString, &cloudrun.SetIamPolicyRequest{
```
Comment (vangent) [inline internal/cmd/gocdk/launch.go]: IIUC, we only do this the first time the service is launched? So if the user changes it, we won't clobber their change?

File: internal/cmd/gocdk/launch.go [hunk 10]
```diff
@@ -187,6 +209,172 @@ func (local *localLauncher) Launch(ctx context.Context, input *LaunchInput) (*ur
...
+	tick := time.NewTicker(5 * time.Second)
+	defer tick.Stop()
+	for {
+		select {
+		case <-tick.C:
+		case <-ctx.Done():
+			return nil, xerrors.Errorf("cloud run launch: wait for ready: %w", ctx.Err())
+		}
+
+		currService, err := crl.client.Projects.Locations.Services.Get("projects/" + projectID + "/locations/" + location + "/services/" + serviceName).Context(ctx).Do()
```
Comment (vangent) [inline internal/cmd/gocdk/launch.go]: Don't we have this as `serviceString` already?

File: internal/cmd/gocdk/launch.go [hunk 11]
```diff
@@ -187,6 +209,172 @@ func (local *localLauncher) Launch(ctx context.Context, input *LaunchInput) (*ur
...
+		if len(out) == 0 {
+			return xerrors.Errorf("docker push: %w", err)
+		}
+		return xerrors.Errorf("docker push:\n%s", out)
+	}
+	return nil
+}
+
+// conditionStatus finds the Cloud Run condition with the given name and returns
+// its status string (one of "True", "False", or "Unknown").
```
Comment (vangent) [inline internal/cmd/gocdk/launch.go]: ... or the empty string, right?

File: internal/cmd/gocdk/launch.go [hunk 12]
```diff
@@ -187,6 +209,172 @@ func (local *localLauncher) Launch(ctx context.Context, input *LaunchInput) (*ur
...
+			if err != nil {
+				return nil, xerrors.Errorf("cloud run launch: parse service URL: %w", err)
+			}
+			return u, nil
+		}
+		crl.logger.Println("Not ready yet, will poll again...")
+	}
+}
+
+func (crl *cloudRunLauncher) push(ctx context.Context, imageName string) error {
```
Comment (vangent) [inline internal/cmd/gocdk/launch.go]: Nit: maybe dockerPush?

File: internal/cmd/gocdk/launch.go [hunk 13]
```diff
@@ -187,6 +209,172 @@ func (local *localLauncher) Launch(ctx context.Context, input *LaunchInput) (*ur
...
+func conditionStatus(conds []*cloudrun.ServiceCondition, name string) string {
+	for _, c := range conds {
+		if c.Type == name {
+			return c.Status
+		}
+	}
+	return ""
+}
+
+// launchSpecifier is the set of arguments passed from a biome's Terraform
```
Comment (vangent) [inline internal/cmd/gocdk/launch.go]: zombie type :-)

File: internal/cmd/gocdk/launch.go (hunk 4 above)
Comment (zombiezen) [reply internal/cmd/gocdk/launch.go]: Done.

File: internal/cmd/gocdk/launch.go (hunk 6 above)
Comment (zombiezen) [reply internal/cmd/gocdk/launch.go]: Done.

File: internal/cmd/gocdk/launch.go (hunk 7 above)
Comment (zombiezen) [reply internal/cmd/gocdk/launch.go]: Done.

File: internal/cmd/gocdk/launch.go (hunk 10 above)
Comment (zombiezen) [reply internal/cmd/gocdk/launch.go]: Yes. Fixed.

File: internal/cmd/gocdk/launch.go (hunk 12 above)
Comment (zombiezen) [reply internal/cmd/gocdk/launch.go]: Done.

File: internal/cmd/gocdk/launch.go (hunk 11 above)
Comment (zombiezen) [reply internal/cmd/gocdk/launch.go]: Done.

File: internal/cmd/gocdk/launch.go (hunk 1 above)
Comment (vangent) [reply internal/cmd/gocdk/launch.go]: SGTM.

File: internal/cmd/gocdk/launch.go (hunk 3 above)
Comment (vangent) [reply internal/cmd/gocdk/launch.go]: SGTM.

File: internal/cmd/gocdk/launch.go (hunk 5 above)
Comment (vangent) [reply internal/cmd/gocdk/launch.go]: OK.
